vault/request_handling_list_filtering.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2025 OpenBao a Series of LF Projects, LLC
+// SPDX-License-Identifier: MPL-2.0
+
+package vault
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/openbao/openbao/helper/identity"
+	"github.com/openbao/openbao/sdk/v2/helper/template"
+	"github.com/openbao/openbao/sdk/v2/logical"
+)
+
+func (c *Core) filterListResponse(ctx context.Context, req *logical.Request, unauth bool, auth *logical.Auth, acl *ACL, te *logical.TokenEntry, entity *identity.Entity, resp *logical.Response) error {
+	// Ignore lists with no results.
+	if resp == nil {
+		return nil
+	}
+
+	// Ignore non-list operations.
+	switch req.Operation {
+	case logical.ListOperation:
+	case logical.ScanOperation:
+	default:
+		return nil
+	}
+
+	// No filtering, so nothing to do.
+	if auth.ResponseKeysFilterPath == "" {
+		return nil
+	}
+
+	// Secret and Auth should not be set on List operation responses which
+	// expect to be filtered.
+	if resp.Secret != nil {
+		c.logger.Error("non-empty secret on filtered list response", "path", req.Path)
+		return ErrInternalError
+	}
+	if resp.Auth != nil {
+		c.logger.Error("non-empty secret on filtered list response", "path", req.Path)
+		return ErrInternalError
+	}
```

Comment:
I'd happily keep them separated if the differed at least in log message (or return value), but in this case:
```suggestion
	if resp.Secret != nil || resp.Auth != nil {
		c.logger.Error("non-empty secret on filtered list response", "path", req.Path)
		return ErrInternalError
	}
```